Additional font tweaks

aaronshekey opened the pull request:
Rolling up any tweaks we may need to the font stack as a hotfix.

These changes can be [previewed on the Stacks documentation](https://deploy-preview-642--stacks.netlify.app/).

aaronshekey committed:
Remove system-ui

This was causing some trouble on Windows when anything other than English was selected. It would default to the International fonts, though, one could argue this was the intent we were after. ¯\_(ツ)_/¯

aaronshekey committed:
Kill Ubuntu

Seems Ubuntu users really don't like the system font provided by Ubuntu

aaronshekey committed:
Remove Roboto

This was causing some wild issues on Debian, and Android will fall through to Roboto anyway

aaronshekey committed:
Update _stacks-constants-type.less

MatthiasPortzel commented:
I think it's worth noting, `ui-monospace` is only supported by Safari [[1]](https://caniuse.com/extended-system-fonts). This means macOS users on browsers other than Safari (the majority) don't get SF Mono. Unless they have another one of these fonts installed, they'll fall-back to Menlo.

aaronshekey commented:
> Unless they have another one of these fonts installed, they'll fall-back to Menlo.

This is what I've found as well. Currently, there doesn't seem to be a way outside of that progressively-enhanced `ui-monospace` CSS property to actually get San Francisco Mono to display. I've spec'd that in the hopes that other browsers on macOS will adopt that. Thanks for calling it out.

LunarWatcher commented:
Please don't give us Arial back everywhere: https://meta.stackexchange.com/questions/364048/we-are-switching-to-system-fonts-on-may-10-2021?cb=1#comment1214996_364048 https://meta.stackexchange.com/questions/364048/we-are-switching-to-system-fonts-on-may-10-2021?cb=1#comment1215982_364097

I'm starting to see the value more and more in just letting the browser determine the font - readable is subjective enough for no single font stack to universally work.

fireattack commented:
@aaronshekey  Can't leave comments in https://meta.stackexchange.com/a/364291/998555 due to reputation, so I think I would leave a comment here.

The reason you see discrepancy between Firefox and Chrome is because Firefox [doesn't support](https://developer.mozilla.org/en-US/docs/Web/CSS/font-family#browser_compatibility) `system-ui` keyword at all. It simply ignores it. 

Also, the poor result in non-English Windows is the exactly reason why GitHub, Twitter, etc. have tried and then immediately reverted their change of font stack of including system-ui.

aaronshekey commented:
> Also, the poor result in non-English Windows is the exactly reason why GitHub, Twitter, etc. have tried and then immediately reverted their change of font stack of including system-ui.

Fair enough! I feel pretty confident leaving this off now.

aaronshekey commented:
> Please don't give us Arial back everywhere

Based on the order of things, this should only affect Linux distros, and lets us maintain the status quo across all Linux installs. One person's "Arial" will be "Liberation Sans" on Ubuntu. It'll be "Roboto" on Android, etc.

aaronshekey commented:
After my testing in Linux distros, here are the following font mappings with and without Arial. Looks like Chrome actually installs Liberation Sans, at least it wanted to on Debian, but I failed to complete the install.

So, the logic in including `Arial` here, is for some normalization across Linux distros between Chrome and Firefox ¯\_(ツ)_/¯ 

| OS | Firefox with Arial | Firefox without Arial | Chrome with Arial | Chrome without Arial |
| --- | --- | --- | --- | --- |
| **Ubuntu** | Liberation Sans | DejaVu Sans | Liberation Sans | Liberation Sans |
| **Fedora** | Liberation Sans | DejaVu Sans | Liberation Sans | Liberation Sans | 

I plan on testing Debian, and CentOS, but I'm struggling to get Chrome installed on Debian, and CentOS doesn't seem to like my network adapter.

aaronshekey committed:
Add Liberation Sans for consistency on Linux distros

aaronshekey committed:
Get the name right, of course

aaronshekey commented:
This branch is now spec'ing "Liberation Sans" and "Liberation Mono" for a wide variety of Linux desktop distros.

Elliot2560 commented:
Actually I think the substitution for Arial on Linux depends on a few other things too. If you have Arimo installed, which you would if you installed all the Google web fonts, or if you have the `fonts-croscore` package on Ubuntu, then Arimo takes precedence. This is my experience on Arch. I'm not entirely sure how it works on Chrome, but it seems to me that Firefox uses the substitution rules set by fontconfig, and the precedence goes `Arimo > Liberation Sans > Albany > Albany MT`, I think. (Check out [this conf file](https://github.com/freedesktop/fontconfig/blob/master/conf.d/30-metric-aliases.conf) where it's specified.)

Though in the end these could all be reasons why it makes (more) sense to specify Liberation Sans and Liberation Mono. I can also confirm that Liberation Sans is a dependency of both Firefox and Chromium.

aaronshekey commented:
@LunarWatcher Care to chime in or are you good with 👎 each commit? 😛

LunarWatcher commented:
I've come to fundamentally dislike Arial and clones for their awful readability when contextual reading fails. That often happens on abbreviations, or more relevant to SO, on certain posts with enough typos for every letter to be in question. (Or just whenever my brain decides to nope out and crash contextual reading) 

That being said, it seems that opinion isn't a mainstream one (people on meta are practically begging to have Arial back), and it's therefore unlikely to lead anywhere. So for now, I'm good with the 👎s and silently disappearing into userstyles and whatever layout breaking that entails.

pxeger commented:
Please just stop trying to be clever. Use `sans-serif`. Just `sans-serif`. It exists for a reason - the browser does it very damn best to make sure it's accessible, quick to load, and looks good. And it's customisable on most browsers without needing unreliable user styles.

fireattack commented:
That's something I never understand about all these "font stacks" either. I honestly feel they're just re-inventing the wheels.

Sure, `sans-serif` may not be the "system fonts" on some OSs by default, but all the modern browsers have the ability to let the user adjust the "default" font, which is exactly what `sans-serif` would choose to use. 

IMO this is just straight up superior than forcing the "system font" using hard-coded rule (if one really want to use system font, they can always set it as their browser's default font. Not to mention, in some OSs like Android/iOS, the default font is already system font).

aaronshekey commented:
As I explained in my [post on meta](https://meta.stackexchange.com/questions/364048/we-are-switching-to-system-fonts-on-may-10-2021):

> Apple’s San Francisco and Microsoft’s Segoe both look great on retina displays, have more expressive weights, and improve readability across all contexts. With more weights, we can have better hierarchy. We can also get more expressive. For example, in dark mode, we could drop our font weights a bit for better readability.

If I choose `sans-serif`, it will render as Helvetica on iOS and macOS. It will render as Arial on Windows. These typefaces have none of those benefits I mentioned, none of the font hinting built into these new typefaces. We'd rather not wait until the operating systems update their underlying definition of `sans-serif`. Additionally, especially on Linux, it varies per browser.

> Accessible, quick to load, and looks good

Yep, these are our goals. This font stack is designed to choose the newest, most accessible _screen native_ typeface as possible. :v:

fireattack commented:
>It will render as Arial on Windows. These typefaces have none of those benefits I mentioned, none of the font hinting built into these new typefaces

Like I said, It will render as Arial because that's the default font of the *browser*: 

Chrome: 

![image](https://user-images.githubusercontent.com/4130991/118389083-e04dbe00-b5ed-11eb-9d9e-bf59839a3b23.png)

Firefox:

![image](https://user-images.githubusercontent.com/4130991/118389148-0e330280-b5ee-11eb-8266-125f59dda2de.png)

It has nothing to do with Windows. These default choices are hard-coded into browser due to historical reasons (there are tickets on both Chromium and Firefox's bug trackers about changing them, but no enough traction yet.)

Users can change it to Segoe if they don't like Arial. 

`Sans-serif` simply uses these settings (same for `serif` and `monospace`). System's built-in font fallback will only be engaged if it can't display certain glyphs. 

 I assume it would be similar for MacOS, but I don't use it. 

You do have a case for iOS though.

Edit: by the way, all these font stacks have the drawback that seldom got talked about: if someone happened to have installed some of these fonts in the rule on their OS, it has the chance to mess up the display even further by showing a font that is not from their OS, which invalidates the "native look" point.

aaronshekey committed:
Remove font smoothing

This is really a tough call. As of macOS 11.3, the user can no longer choose their antialiasing at the OS level. This _feels_ like a bug or oversight, so the idea was we would be opinionated since the users can't make a change on their end anymore. Not everyone is on 11.3, and I hope this is fixed in 11.4, so let's revisit then.

aaronshekey commented:
I hear you on browser-default fonts. But remember, we want to modernize our fonts without too much user intervention and feel a bit more native to their OS.

stamminator commented:
What I hear from @aaronshekey is that the hesitation of browser vendors to default their browsers to more modern fonts (with all the advantages they entail) should not prevent SO from benefitting from those fonts in as "native", per-platform a way as possible. Provided SO can work out any kinks in a timely manner, I don't see why this is a bad take.
